src/app-layer-dnp3.c

```diff
@@ -840,6 +840,9 @@ static void DNP3HandleUserDataRequest(DNP3State *dnp3, const uint8_t *input,
 
     lh = (DNP3LinkHeader *)input;
 
+    if (input_len <= sizeof(DNP3LinkHeader))
```

Comment:
No, it's not tested there since input_len is tested at 1060 but frame_len is passed to DNP3HandleUserDataRequest.
Well, probably frame_len is tested inside DNP3CalculateLinkLength so the test at 843 is not necessary.
Please see my reply to your second note with a complete explanation.

Reply:
Indeed but line 1067 adds test `if (input_len < frame_len) {` after `if (input_len < sizeof(DNP3LinkHeader)) {`
So, this line 843 test looks redundant to me.

Reply:
line 1067 forces waiting for the data if frame_len is bigger than input_len.
However, line 843 tests quite different condition, i.e. if frame_len is small (smaller than header length).
Well, I agree that 843 is redundant. But it's only because of DNP3CalculateLinkLength never returns value lesser than header length (or zero).
There is no other validation and 1067 would not do the trick if DNP3CalculateLinkLength could return any value.

Reply:
Ok, I understand better what you mean.
`DNP3CalculateLinkLength` is indeed the good function checking that we have enough space in the frame (to use the name used in the source), even if we have more frames in the buffer.

One can use `DEBUG_VALIDATE_BUG_ON` to add these safety checks, even if it is redundant today